src/unix/apple/macos/disk.rs

```diff
@@ -91,36 +71,91 @@ pub(crate) fn get_disk_type(disk: &libc::statfs) -> Option<DiskKind> {
             let properties_result = unsafe {
                 CFReleaser::new(ffi::IORegistryEntryCreateCFProperty(
                     current_service_entry.inner(),
-                    characteristics_string.inner(),
+                    key.inner(),
                     kCFAllocatorDefault,
                     0,
                 ))
             };
 
-            if let Some(device_properties) = properties_result {
-                let disk_type = unsafe {
-                    super::disk::get_str_value(
-                        device_properties.inner(),
-                        DictKey::Defined(ffi::kIOPropertyMediumTypeKey),
-                    )
-                };
-
-                if let Some(disk_type) = disk_type.and_then(|medium| match medium.as_str() {
-                    _ if medium == ffi::kIOPropertyMediumTypeSolidStateKey => Some(DiskKind::SSD),
-                    _ if medium == ffi::kIOPropertyMediumTypeRotationalKey => Some(DiskKind::HDD),
-                    _ => None,
-                }) {
-                    return Some(disk_type);
-                } else {
-                    // Many external drive vendors do not advertise their device's storage medium.
-                    //
-                    // In these cases, assuming that there were _any_ properties about them registered, we fallback
-                    // to `HDD` when no storage medium is provided by the device instead of `Unknown`.
-                    return Some(DiskKind::HDD);
-                }
+            if let Some(result) =
+                properties_result.and_then(|properties| eval(parent_entry, &properties))
+            {
+                return Some(result);
             }
         }
     }
 
     None
 }
+
+pub(crate) fn get_disk_type(bsd_name: &[u8]) -> Option<DiskKind> {
+    let characteristics_string = unsafe {
+        CFReleaser::new(cfs::CFStringCreateWithBytesNoCopy(
+            kCFAllocatorDefault,
+            ffi::kIOPropertyDeviceCharacteristicsKey.as_ptr(),
+            ffi::kIOPropertyDeviceCharacteristicsKey.len() as _,
+            cfs::kCFStringEncodingUTF8,
+            false as _,
+            kCFAllocatorNull,
+        ))?
+    };
+
+    iterate_service_tree(bsd_name, characteristics_string, |_, properties| {
+        let disk_type = unsafe {
+            super::disk::get_str_value(
+                properties.inner(),
+                DictKey::Defined(ffi::kIOPropertyMediumTypeKey),
+            )
+        };
+
+        if let Some(disk_type) = disk_type.and_then(|medium| match medium.as_str() {
+            _ if medium == ffi::kIOPropertyMediumTypeSolidStateKey => Some(DiskKind::SSD),
+            _ if medium == ffi::kIOPropertyMediumTypeRotationalKey => Some(DiskKind::HDD),
+            _ => None,
+        }) {
+            Some(disk_type)
+        } else {
+            Some(DiskKind::Unknown(-1))
+        }
```

Comment:
What do you think about:

```suggestion
        let disk_type = unsafe {
            super::disk::get_str_value(
                properties.inner(),
                DictKey::Defined(ffi::kIOPropertyMediumTypeKey),
            )
        }?;

        match medium.as_str() {
            _ if medium == ffi::kIOPropertyMediumTypeSolidStateKey => Some(DiskKind::SSD),
            _ if medium == ffi::kIOPropertyMediumTypeRotationalKey => Some(DiskKind::HDD),
            _ => Some(DiskKind::Unknown(-1)),
        }
```

Reply:
Much easier to reason with, thanks. I will update!